orrery/include/orrery/component.h

```diff
@@ -0,0 +1,69 @@
+// Concord
+//
+// Copyright (c) 2021 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the
+// "License").  You may not use this product except in compliance with the
+// Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#pragma once
+
+#include <utility>
+#include <type_traits>
+
+#include "kvstream.h"
+#include "orrery_msgs.cmf.hpp"
+#include "Logger.hpp"
+#include <iostream>
+
+namespace concord::orrery {
+
+// A component provides an interface to handle `AllMsgs`
+class IComponent {
+ public:
+  virtual ~IComponent(){};
+  virtual void handle(ComponentId from, AllMsgs&&) = 0;
+};
+
+template <typename ComponentImpl>
+class Component : public IComponent {
+ public:
+  // Trait to ensure that a ComponentImpl has a `handle` method that takes a message of type `Msg`.
+  template <typename Impl, typename Msg, typename = std::void_t<>>
+  struct CanHandleMsgT : std::false_type {};
+
+  template <typename Impl, typename Msg>
+  struct CanHandleMsgT<
+      Impl,
+      Msg,
+      std::void_t<decltype(std::declval<Impl>().handle(std::declval<ComponentId>(), std::declval<Msg>()))>>
+      : std::true_type {};
```

Comment:
I was wondering if the `CanHandleMsg` type should be out of `Component` or we should drop the `Impl` template parameter. I guess we can drop it as every `Component` for a specific `ComponentImpl` will have its own `CanHandleMsg`.

I only thought about it, because I wanted to find out if `Impl` is the same as `ComponentImpl`. What do you think?

Reply:
Ah yes, I think you are right and we can drop it. IIRC `Impl` should be the same as `ComponentImpl`.

The more important thing that I didn't implement and I think is important is to allow nested visitation for subvariants. That would be super powerful. There's a comment in one of the examples.